## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [1]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [2]:
# Write your code for importing the csv file here
movies = pd.read_csv("Movies.csv")
movies.head(10)

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   
5  Color          Sam Raimi                   392.0     156.0   
6  Color       Nathan Greno                   324.0     100.0   
7  Color        Joss Whedon                   635.0     141.0   
8  Color        David Yates                   375.0     153.0   
9  Color        Zack Snyder                   673.0     183.0   

   director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                      0.0                   855.0   Joel David Moore   
1                    563.0                  1000.0      Orlando Bloom   
2                      0.0                   161.0       Rory Kinnear   
3                  22000.0                 23000.0     Christian Bale   
4                    475.0                   530.0    Samantha Morton   
5                      0.0                  4000.0       James Franco   
6                     15.0                   284.0       Donna Murphy   
7                      0.0                 19000.0  Robert Downey Jr.   
8                    282.0                 10000.0   Daniel Radcliffe   
9                      0.0                  2000.0       Lauren Cohan   

   actor_1_facebook_likes        gross  \
0                  1000.0  760505847.0   
1                 40000.0  309404152.0   
2                 11000.0  200074175.0   
3                 27000.0  448130642.0   
4                   640.0   73058679.0   
5                 24000.0  336530303.0   
6                   799.0  200807262.0   
7                 26000.0  458991599.0   
8                 25000.0  301956980.0   
9                 15000.0  330249062.0   

                                              genres  ...  \
0                    Action|Adventure|Fantasy|Sci-Fi  ...   
1                           Action|Adventure|Fantasy  ...   
2                          Action|Adventure|Thriller  ...   
3                                    Action|Thriller  ...   
4                            Action|Adventure|Sci-Fi  ...   
5                           Action|Adventure|Romance  ...   
6  Adventure|Animation|Comedy|Family|Fantasy|Musi...  ...   
7                            Action|Adventure|Sci-Fi  ...   
8                   Adventure|Family|Fantasy|Mystery  ...   
9                            Action|Adventure|Sci-Fi  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                738.0  English      USA           PG-13  263700000.0   
5               1902.0  English      USA           PG-13  258000000.0   
6                387.0  English      USA              PG  260000000.0   
7               1117.0  English      USA           PG-13  250000000.0   
8                973.0  English       UK              PG  250000000.0   
9               3018.0  English      USA           PG-13  250000000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4      2012.0                  632.0        6.6          2.35   
5      2007.0                11000.0        6.2          2.35   
6      2010.0                  553.0        7.8          1.85   
7      2015.0           

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [3]:
# Write your code for inspection here
#movies.shape
#movies.info()
#movies.describe()
null_columns=movies.columns[movies.isnull().any()]
null_columns
hh=pd.DataFrame(movies[null_columns].isnull().sum())
hh.size
#print(movies[movies.isnull().any(axis=1)].value_counts())
print((movies.isnull().sum()>0).sum())


12


#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [4]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations

movies.drop(columns=['color','director_facebook_likes','actor_1_facebook_likes','actor_2_facebook_likes','actor_3_facebook_likes','actor_2_name','cast_total_facebook_likes','actor_3_name','duration','facenumber_in_poster','content_rating','country','movie_imdb_link','aspect_ratio','plot_keywords'] ,axis=1,inplace=True)
movies.head(10)
movies.shape

(3853, 13)

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [5]:
# Write you code here
null_columns=movies.columns[movies.isnull().any()]
null_columns
hh=pd.DataFrame(movies[null_columns].isnull().sum())
hh
round(100*(movies.isnull().sum()/len(movies.index)), 2)

director_name             0.00
num_critic_for_reviews    0.03
gross                     0.00
genres                    0.00
actor_1_name              0.00
movie_title               0.00
num_voted_users           0.00
num_user_for_reviews      0.00
language                  0.08
budget                    0.00
title_year                0.00
imdb_score                0.00
movie_facebook_likes      0.00
dtype: float64

#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [6]:
# Write your code for filling the NaN values in the 'language' column here
values = {'language': 'English'}
movies.fillna(value=values,inplace=True)
len(movies[movies["language"]=='English'])
(movies.language == 'English').sum()

3674

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [7]:
# Write your code for unit conversion here

movies["budget"]=movies["budget"].apply(lambda x : x/1000000)
movies["gross"]=movies["gross"].apply(lambda x : x/1000000)
# round to two decimal places in python pandas 
#pd.options.display.float_format = '{:.2f}'.format
movies.head(5)

director_name  num_critic_for_reviews       gross  \
0      James Cameron                   723.0  760.505847   
1     Gore Verbinski                   302.0  309.404152   
2         Sam Mendes                   602.0  200.074175   
3  Christopher Nolan                   813.0  448.130642   
4     Andrew Stanton                   462.0   73.058679   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
4          Action|Adventure|Sci-Fi     Daryl Sabara   

                                 movie_title  num_voted_users  \
0                                    Avatar            886204   
1  Pirates of the Caribbean: At World's End            471220   
2                                   Spectre            275868   
3                     The Dark Knight Rises           1144337   
4                               John Carter            212204   

   num_user_for_reviews language  budget  title_year  imdb_score  \
0                3054.0  English   237.0      2009.0         7.9   
1                1238.0  English   300.0      2007.0         7.1   
2                 994.0  English   245.0      2015.0         6.8   
3                2701.0  English   250.0      2012.0         8.5   
4                 738.0  English   263.7      2012.0         6.6   

   movie_facebook_likes  
0                 33000  
1                     0  
2                 85000  
3                164000  
4                 24000

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [8]:
# Write your code for creating the profit column here
movies["profit"]=(movies["gross"]-movies["budget"])
movies.head(5)

director_name  num_critic_for_reviews       gross  \
0      James Cameron                   723.0  760.505847   
1     Gore Verbinski                   302.0  309.404152   
2         Sam Mendes                   602.0  200.074175   
3  Christopher Nolan                   813.0  448.130642   
4     Andrew Stanton                   462.0   73.058679   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
4          Action|Adventure|Sci-Fi     Daryl Sabara   

                                 movie_title  num_voted_users  \
0                                    Avatar            886204   
1  Pirates of the Caribbean: At World's End            471220   
2                                   Spectre            275868   
3                     The Dark Knight Rises           1144337   
4                               John Carter            212204   

   num_user_for_reviews language  budget  title_year  imdb_score  \
0                3054.0  English   237.0      2009.0         7.9   
1                1238.0  English   300.0      2007.0         7.1   
2                 994.0  English   245.0      2015.0         6.8   
3                2701.0  English   250.0      2012.0         8.5   
4                 738.0  English   263.7      2012.0         6.6   

   movie_facebook_likes      profit  
0                 33000  523.505847  
1                     0    9.404152  
2                 85000  -44.925825  
3                164000  198.130642  
4                 24000 -190.641321

In [9]:
# Write your code for sorting the dataframe here
top10=movies.sort_values(by=['profit'],ascending=False)
top10.head(5)

director_name  num_critic_for_reviews       gross  \
0        James Cameron                   723.0  760.505847   
28     Colin Trevorrow                   644.0  652.177271   
25       James Cameron                   315.0  658.672302   
2704      George Lucas                   282.0  460.935665   
2748  Steven Spielberg                   215.0  434.949459   

                                genres         actor_1_name  \
0      Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
28    Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
25                       Drama|Romance    Leonardo DiCaprio   
2704   Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
2748                     Family|Sci-Fi         Henry Thomas   

                              movie_title  num_voted_users  \
0                                 Avatar            886204   
28                        Jurassic World            418214   
25                               Titanic            793059   
2704  Star Wars: Episode IV - A New Hope            911097   
2748          E.T. the Extra-Terrestrial            281842   

      num_user_for_reviews language  budget  title_year  imdb_score  \
0                   3054.0  English   237.0      2009.0         7.9   
28                  1290.0  English   150.0      2015.0         7.0   
25                  2528.0  English   200.0      1997.0         7.7   
2704                1470.0  English    11.0      1977.0         8.7   
2748                 515.0  English    10.5      1982.0         7.9   

      movie_facebook_likes      profit  
0                    33000  523.505847  
28                  150000  502.177271  
25                   26000  458.672302  
2704                 33000  449.935665  
2748                 34000  424.449459

In [10]:
top10 = # Write your code to get the top 10 profiting movies here

SyntaxError: invalid syntax (<ipython-input-10-8d1b1732725d>, line 1)

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [ ]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'
IMDb_Top_250=movies.sort_values(by=['imdb_score'],ascending=False).head(250)
IMDb_Top_250[IMDb_Top_250.num_voted_users > 25000]
IMDb_Top_250.head(5)


In [ ]:
df=IMDb_Top_250.copy()

df["group"]=pd.cut(df["imdb_score"], np.arange(7.5,10.5, 0.5))
df.groupby(['group']).size()

In [ ]:
IMDb_Top_250 = movies.sort_values(by = 'imdb_score', ascending = False)
IMDb_Top_250 = IMDb_Top_250.loc[IMDb_Top_250.num_voted_users > 25000]
IMDb_Top_250 = IMDb_Top_250.iloc[:250, ]
IMDb_Top_250['Rank'] = range(1,251)

import matplotlib.pyplot as plt
plt.hist(IMDb_Top_250['imdb_score'], bins = 5, range = (7.5,10), edgecolor = 'cyan')
plt.show()

#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [12]:
# Write your code for creating three new dataframes here
Meryl_Streep = movies[movies["actor_1_name"] =="Meryl Streep"]
Meryl_Streep.size                


154

In [13]:
Leo_Caprio = movies[movies["actor_1_name"].str.contains('Leonardo DiCaprio')]
Leo_Caprio.size

294

In [14]:
Brad_Pitt = movies[movies["actor_1_name"].str.contains('Brad Pitt')]
Brad_Pitt.size

238

In [17]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep,Leo_Caprio,Brad_Pitt],axis=0)
Combined.head(5)

director_name  num_critic_for_reviews       gross  \
392     Nancy Meyers                   187.0  112.703470   
1038   Curtis Hanson                    42.0   46.815748   
1132     Nora Ephron                   252.0   94.125426   
1322   David Frankel                   208.0  124.732962   
1390  Robert Redford                   227.0   14.998070   

                               genres  actor_1_name             movie_title  \
392              Comedy|Drama|Romance  Meryl Streep       It's Complicated    
1038  Action|Adventure|Crime|Thriller  Meryl Streep         The River Wild    
1132          Biography|Drama|Romance  Meryl Streep          Julie & Julia    
1322             Comedy|Drama|Romance  Meryl Streep  The Devil Wears Prada    
1390               Drama|Thriller|War  Meryl Streep        Lions for Lambs    

      num_voted_users  num_user_for_reviews language  budget  title_year  \
392             69860                 214.0  English    85.0      2009.0   
1038            32544                  69.0  English    45.0      1994.0   
1132            79264                 277.0  English    40.0      2009.0   
1322           286178                 631.0  English    35.0      2006.0   
1390            41170                 298.0  English    35.0      2007.0   

      imdb_score  movie_facebook_likes     profit  
392          6.6                     0  27.703470  
1038         6.3                     0   1.815748  
1132         7.0                 13000  54.125426  
1322         6.8                     0  89.732962  
1390         6.2                     0 -20.001930

In [27]:
# Write your code for grouping the combined dataframe here
Combined_by_segment=Combined.groupby('actor_1_name')
Combined_by_segment['num_user_for_reviews'].mean()

actor_1_name
Brad Pitt            742.352941
Leonardo DiCaprio    914.476190
Meryl Streep         297.181818
Name: num_user_for_reviews, dtype: float64

In [28]:
# Write the code for finding the mean of critic reviews and audience reviews here
Combined_by_segment['num_critic_for_reviews'].mean()


actor_1_name
Brad Pitt            245.000000
Leonardo DiCaprio    330.190476
Meryl Streep         181.454545
Name: num_critic_for_reviews, dtype: float64

#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>